ECDH encryption/decryption support

zapu opened the pull request:
This pull request introduces encryption and decryption using ECDH keys (no cv25519 yet though, read below). Most of new code is in `openpgp/packet/ecdh.go` and `openpgp/ecdh/ecdh.go`. Package `ecdh` got all the functions that were independent enough not to need to import anything from package `packet` (which would create circular dependency) - random scalars, KDF (but creating "KDFparams" buffer is in `packet`),  rfc3394 key wrapping, padding. 

Encryption is done mostly in `ecdh` package, but a lot of things have to be passed as arguments. Decryption is mostly in `packet` package (`ecdh.go` file). Let me know which is better - I'll either rewrite Decryption in similar fashion, or move encryption to the `packet` package.

There are new functions for reading MPI coords for ECDH in `openpgp/packet/packet.go`. `EncryptedKey` has new `ecdh_C []byte` field, because ECDH encryption serializes curve coordinates and AES-keywrapped buffer.

`ecdh_test.go` now also tests encryption/decryption. The test that was already there was "completed" to decrypt message, and also there is a new test that does roundtrip encryption/decryption with test additional two keys with two different curves, created with gpg.

CV25519 needs EdDSA support. There is some ground work being done, but it's very far from being able to actually sign anything. Also there is no ECDH key generation - I haven't gotten that far in understanding go-crypto yet.

The question is, where to go with this. I've found a fork [1] that also implements ECDH and has better go code that I was able to write. It also does some much needed cleanup in other places. I can try to bring that code in, instead. I wish I had found it sooner :wink: 

[1] https://github.com/benburkert/openpgp

zapu committed:
Encryption and decryption with ECDH keys

zapu commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+	if S_len > len(S) {
+		// If we got invalid S (bigger than curve's P), we are going
+		// to produce invalid key. Garbage in, garbage out.
+		buf.Write(make([]byte, S_len-len(S)))
```
What's the best way to write *n* zero bytes somewhere?

maxtaco replied:
This is a fine way, assuming `S_len-len(S)` isn't huge.  Otherwise, it would be do so repeatedly.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+// nonZeroRandomBytes fills the given slice with non-zero random octets.
+func nonZeroRandomBytes(s []byte, rand io.Reader) (err error) {
```
This shouldn't be necessary.  See docs on rand.Read [here](https://golang.org/src/crypto/rand/rand.go?s=761:799#L12).

zapu replied:
Oh, I was going to ask about it, thanks for noticing. I took it from https://github.com/keybase/go-crypto/blob/master/openpgp/elgamal/elgamal.go#L106 but there is also slightly different version here: https://github.com/keybase/go-crypto/blob/master/rsa/pkcs1v15.go#L176

I can try to go with what KBPGP does, which I think keeps pulling random bignums with bit length = N.bitLen until it finds one between 1 and N (and it usually just finds it right away).

maxtaco replied:
I misunderstood the intention here. What you have is great, ignore my comment above.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+	return m, nil
+}
+
+func (e *PublicKey) Kdf(S []byte, kdf_params []byte, hash crypto.Hash) []byte {
```
I believe proper Go style would have you call this function `KDF`.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+
+	return key
+}
+
```
provide some commentary on which RFC/section is being implemented.

maxtaco replied:
I recognize this code! Is it transposed from `rfc3394.iced`?  If so, please provide a comment. In case there are bugs, we can upstream any patches.

maxtaco replied:
This function should be called `AESKeyUnwrap`

zapu replied:
Oh, there is a comment about RFC but in the next function, I think I must have reordered them. But I will leave more comments about this.

It's a little bit different than what KBPGP does, as it does not split the buffer into 64bit words first, it operates on the byte buffer. But it's essentially the non-shifting algorithm from the RFC.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+// Implements RFC 3394 Key Wrapping.
+func AesKeyWrap(key, plainText []byte) ([]byte, error) {
```
Rename to `AESKeyWrap`

maxtaco commented on openpgp/ecdh_test.go:
```diff
@@ -91,15 +87,121 @@ func openAndDecryptKey(t *testing.T, key string, passphrase string) EntityList {
+3UeahfoOyLOY71uHDNdfB66DCa0BANK4aMk+j7bpJoWFNpkWq9JnhpfXV9L2dh3R
+kKuKwZkI
+=RN+z
+-----END PGP PRIVATE KEY BLOCK-----`
```
can you test some failure cases?

zapu replied:
Will do! Thanks

zapu replied:
So we mostly want to test that invalid payloads fail to decrypt with error, instead of panic-ing the whole program, right?

maxtaco replied:
Exactly.  What happens when you decrypt forged or corrupted encryptions?  Or if you try to decrypt with the wrong secret key.

maxtaco commented on openpgp/packet/ecdh.go:
```diff
@@ -1,18 +1,83 @@
+
+	Sx, err := ecdhpriv.DecryptShared(X, Y)
+	if err != nil {
+		return
```
Less bug-prone to be explicit in your `return`s.  I.e., `return nil, err`.

zapu replied:
Thanks! I've seen the "pascal-like" return value assignment in this repo a few times (I think?) so I kind of followed, but I'm fine with both.

maxtaco commented on openpgp/packet/ecdh.go:
```diff
@@ -1,18 +1,83 @@
+	key := ecdhpriv.Kdf(Sx, kdf_params, hash)
+	decrypted, err := ecdh.AesKeyUnwrap(key[:32], C)
+	if err != nil {
+		return
```
same as above.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+
+	v, err := RandomScalar(random, curve_params.N)
+	if err != nil {
+		return
```
return explicit error here.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+	encrypted, err := AesKeyWrap(key[:32], plain)
+
+	C = encrypted
+	return
```
return explicit values here.

maxtaco reviewed (changes requested):
Made a first pass, I'll make another. Thanks for the great work!

zapu committed:
Style fixes

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+func nonZeroRandomBytes(s []byte, rand io.Reader) (err error) {
+	_, err = io.ReadFull(rand, s)
+	if err != nil {
+		return
```
`return err`

zapu replied:
Apologies for not checking and fixing other occurrences of that `return` form. I'll get to it.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+		for s[i] == 0 {
+			_, err = io.ReadFull(rand, s[i:i+1])
+			if err != nil {
+				return
```
`return err`

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,240 @@
+		}
+	}
+
+	return
```
`return nil`

maxtaco reviewed (commented):
Some more comments...

zapu commented on openpgp/packet/packet.go:
```diff
@@ -525,6 +526,110 @@ func writeMPI(w io.Writer, bitLength uint16, mpiBytes []byte) (err error) {
+		X.bitLength = uint16(coordLen * 8)
+		_, err = readFull(r, X.bytes)
+		if err != nil {
+			return
```
Should I change these as well? It becomes kind of messy with `return nil, nil, err` to return an error.

maxtaco replied:
Yeah, maybe as well change them as well.  I know it's messed, but that's why Go style seems to want.

zapu committed:
More error handling in new MPI io functions

zapu committed:
Add payload with invalid padding to ecdh test

zapu committed:
Return statement style change

maxtaco commented:
I didn't know about @benburkert's fork either.  It might be hard to switch to his fork, since much of the work we have here is be more lenient with weird keys and weird signatures that we have seen in the wild.  However, the GNU extensions like smart card support seem very useful. Probably the best bet is to go forward with the current plan.

In what ways do you think that the other fork is better code than what you have here, and where is the cleanup?

Many thanks

zapu commented:
I like this change: https://github.com/benburkert/crypto/commit/da0a1e107491813ef0ae4029cfe03e1c1844b6a8

But with ECDH it seems like it assumes all encrypted key will consist of encoded X coord and AES ciphertext, which is only correct for cv25519 keys. I have not actually ran their code though, maybe I missed something.

I don't think switching is an option. But I'd like to look at their code while implementing rest of eddsa/cv25510, if it's fine from legal point of view.

maxtaco commented:
I read through those changes too, they are very well done.  My understanding is that the fork inherits the [original license from go-crypto](https://github.com/keybase/go-crypto/blob/master/LICENSE), which is permissive.  

If you want to try to cherry-pick some of these changes, that might be very nice.  I think they are orthogonal, largely, to the stuff that we've done to be more permissive of whacky PGP keys.

zapu committed:
Add another "bad padding" ecdh test

This time one that also fails in gpg2.

zapu committed:
Add invalid AES key test

zapu committed:
Remove error return value from ecdh DecryptShared

It only does curve scalar multiplication, it never fails (in any way
that can be handled).

zapu commented:
I'm going to try to add missing EdDSA support and get cv25519 working now.

zapu committed:
Fix trimming input to AES key wrapping

Hashed data for AES key wrapping should be trimmed to the length equal
to key's cipher key size.

zapu committed:
Fix randomScalar generating secrets 1-byte short

zapu committed:
Add Curve 25519 support for ECDH

zapu commented:
Curve 25519 support is now in. For this I needed to add a `cv25519Curve` type that would conform to curve interface in `elliptic` package. I kind of copied what https://golang.org/src/crypto/elliptic/p224.go to implement specific curves.

After adding the curve, everything more or less worked, but it exposed a bug in key wrapping that I had to fix ("fix input trimming" commit). There was also an issue with randomScalar byte length calculation which would take a few bits away from the random output. I changed it but I'm still not sure if it's correct. But I plan to swap the entire key generation with the one in `elliptic` package anyway - while working on 25519 I discovered that they have key generation along with generating random scalars already.

zapu commented on curve25519/curve_impl.go:
```diff
@@ -0,0 +1,117 @@
+
+func initCv25519() {
+	cv25519.CurveParams = &elliptic.CurveParams{Name: "Curve 25519"}
+	cv25519.P, _ = new (big.Int).SetString("7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed", 16)
```
They are not used by curve operations, but some code depends on getting the curve params to discover curve bit length.

zapu commented:
Also I'm not sure about EdDSA and how much of it is implemented. I was actually surprised that I got cv25519 to work without more EdDSA changes - it was either in much more complete state than I thought or the code does not verify signatures of ecdh keys while reading them. I will look into it more to verify what's left to be done there.

zapu committed:
Swap key generation to the one from elliptic package

Allowed us to remove RandomScalar and rely on elliptic to generate valid
secrets for the curves instead.

zapu committed:
Add tests for invalid cv25519 ecdh payloads

zapu committed:
Add tests for importing cv25519 keys with bad signatures

zapu commented:
I verified that the eddsa key signatures work as well.

@maxtaco Any way we can move forward with this? It's been quite a long time now :)

maxtaco commented on curve25519/curve_impl.go:
```diff
@@ -0,0 +1,117 @@
+// and if so, use 0x40 compressed type to (un)marshal. Otherwise,
+// elliptic.(Un)marshal will be called.
+
+func Marshal(curve elliptic.Curve, x, y *big.Int) []byte {
```
Can you pop this function of out this library and into packet/ecdh.go?  It's confusing to go into `curve25519` to Marshal/Unmarshal NIST curves.

zapu replied:
Hmm, there is slight problem, though. I do the Marshaling/Unmarshaling by testing whether given curve is a 25519 or not. If it is, I do the 0x40 "compressed" format; if it's not, I call Marshal/Unmarshal from `elliptic` package. But I also want to keep `cv25519Curve` type unexported. By doing that, I can't have type cast in other package than `go-crypto/curve25519`.

Too bad Marshal/Unmarshal are not CurveParams methods originaly in `crypto`, this way I could have just overridden them like I do with ScalarMult and others...

zapu replied:
I guess I can move the type cast to curve25519 package, to some "IsCv25519" function. It seems weird, but 25519 is a special case anyway, with it being different curve type than any other that is supported there.

zapu replied:
So this: https://gist.github.com/anonymous/94ee138cadd93978211dffc80f36f5bd seems to work. What do I think? Should I apply this?

(sorry for anonymous gist: I was not logged in on my VM)

maxtaco replied:
I think that's much better, thank you.

maxtaco commented on curve25519/curve_impl.go:
```diff
@@ -0,0 +1,117 @@
+	}
+}
+
+func Unmarshal(curve elliptic.Curve, data []byte) (x, y *big.Int) {
```
Same as above, can you move this into packet/ecdh.go?

maxtaco commented on curve25519/curve_impl.go:
```diff
@@ -0,0 +1,117 @@
+
+var initonce sync.Once
+
+func Cv25519() elliptic.Curve {
```
can the [go `init()`](https://golang.org/doc/effective_go.html#init) function work here?

zapu replied:
I kind of followed what https://golang.org/src/crypto/elliptic/elliptic.go does, but I'll read on the init function and I can try to do that.

maxtaco replied:
Ah I see.  OK, that's a fine thing to emulate. No reason to change it.

maxtaco reviewed (commented):
got started, but more to go.

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+func (e *PublicKey) KDF(S []byte, kdf_params []byte, hash crypto.Hash) []byte {
+	S_len := (e.Curve.Params().P.BitLen() + 7) / 8
```
In general, Go doesn't like snake-cased variables.  Can you rename this to `sLen` or some such?  Otherwise linters will start yelling at us.  Actually, it might make sense to lint all of this code with `golint`.

zapu replied:
Thank you. Sorry for wasting your time with such trivial issues. I'll try to get a linter running.

maxtaco commented on openpgp/packet/packet.go:
```diff
@@ -525,6 +526,83 @@ func writeMPI(w io.Writer, bitLength uint16, mpiBytes []byte) (err error) {
+			return nil, nil, err
+		}
+	} else if header == 0x40 {
+		//if mpiBitSize != coordBitLen + 7 {
```
What's going on with this code?

zapu replied:
oops, I forgot about this. Going to check what's going on there, I commented that out for some reason and it always passes nil as the curve argument there.

zapu replied:
I think it's because when the point is read during EncryptedKey parsing, we don't have our curve yet because we've yet to find our ecdh PrivateKey for that EncryptedKey: https://github.com/zapu/go-crypto/blob/ecdh-rls2/openpgp/packet/encrypted_key.go#L53

They coord bit length could be checked afterwards somewhere, when the key and curve are known.

maxtaco replied:
Without any checks, what would happen if a coordinate is read which is outside of acceptable bit length parameters?  Does the code crash?

maxtaco replied:
BTW, go ahead and delete the commented-out code.

zapu committed:
Move curve point (un)marshaling to ecdh package

zapu committed:
Renamed S_len variable

zapu committed:
Remove underscore variable names from ecdh_test

zapu commented:
I've ran `golint` and fixed what was "mine". It complains a bit about other code, though.

maxtaco commented on curve25519/curve_impl.go:
```diff
@@ -0,0 +1,99 @@
+	*elliptic.CurveParams
+}
+
+func copy_reverse(dst []byte, src []byte) {
```
Call this `copyReverse`

zapu replied:
I've missed this entire file, thanks. I'm doing another run of style-fixing.

maxtaco commented on openpgp/packet/packet.go:
```diff
@@ -525,6 +526,83 @@ func writeMPI(w io.Writer, bitLength uint16, mpiBytes []byte) (err error) {
+	r.Read(buf[:1])
+	header := buf[0]
+	if header == 0x4 {
+		//if mpiBitSize != 2 * coordBitLen + 3 {
```
commented-out code here too.

maxtaco commented on openpgp/packet/public_key.go:
```diff
@@ -130,6 +134,33 @@ func (f *ecdsaKey) newECDSA() (*ecdsa.PublicKey, error) {
 
+func (f *ecdsaKey) newECDH() (*ecdh.PublicKey, error) {
+	var c elliptic.Curve
+	if bytes.Equal(f.oid, oidCurveP256) {
```
Can this be a `switch` statement? It would be cleaner.

zapu committed:
Style fixes

zapu committed:
Check if ECDH/ECDSA public key points lie on curves

zapu committed:
Removed unused nonZeroRandomBytes from ecdh.go

zapu committed:
Nevermind about point checking in public_key.go

zapu committed:
Add getCurveByOid function instead of two if-chains

maxtaco commented:
Looking really good -- can you eliminate any commented-out code?

zapu commented:
I will - I want to add a test case for invalid coordinates in encrypted key.

zapu committed:
Clean readPointMPI up: curve parameter is not needed

Encrypted key point is read with readPointMPI during parsing, while we
don't know the private key yet.

zapu committed:
Correct error handling in readPointMPI

zapu committed:
Add test for readPointMPI different coord len

zapu commented:
Researched and added a test regarding the mpi length checking. GPG doesn't care and neither should we - different implementation may round the number of bits there or not, and both are fine. I created an encrypted message where the coordinates serialized are unnecessarily big, where the bit size is not even different by rounding margin, and they still work fine in GPG.

zapu commented:
Aaaand it fails in KBPGP
```
Error: Need 1059 bits for this curve; got 1459
  at Curve.exports.Curve.Curve._mpi_point_from_slicer_buffer (/home/zapu/pgp/kbpgp/lib/ecc/curves.js:64:15)
```

again, this weird payload works fine in GPG (notice the `AAAAAAAAAAAAAAAAAA...` in armored msg) :
```
~/pgp » gpg2
gpg: Go ahead and type your message ...
-----BEGIN PGP MESSAGE-----
Version: Keybase OpenPGP v2.0.58
Comment: https://keybase.io/crypto

wcA0A6pi0WoSlxTXEgWzBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/ZTuFZa3
go5bAv8SLZd5vTQzjQiqiXfaQUX3dQu+zytgEeiugIshlJ7JykTPGhdQFVSiKQYe
a4RKpgbn8SkNhyIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaR1kL+5vnf9E0wv
vGDKg1jo0uEGYVodxTwk8QuqbxWiCT/jOH9kybjECSPlkDkbUIOezOfaTlwde0Wo
XUNWZ/WrMOJHerpQvMvPNjEwMSGnsIZPh8/Hafj7j8OauMG5EWgCrzsxv1mgsRXP
QkNog/5dM9JIAcT8RpDaFecdhRag6ZPuRKmNuhiFtR7o0spcqX2UkJ3FPB7UydX3
ch9PkTNL1BVD++JqYQE9eaIqlCTAsHwCgO6pQkQqPUvB
=y7cW
-----END PGP MESSAGE-----

gpg: encrypted with 521-bit ECDH key, ID AA62D16A129714D7, created 2016-10-15
      "nist key tester <m+testing@zapu.net>"
purpleschala
^C
gpg: signal Interrupt caught ... exiting

```

maxtaco commented:
you know i recently got a complaint about something similar on EdDSA in
kbpgp.  We can change that check if needs be

On Thu, Jan 26, 2017 at 9:21 AM, Michał Zochniak <notifications@github.com>
wrote:

> Aaaand it fails in KBPGP
>
> Error: Need 1059 bits for this curve; got 1459
>   at Curve.exports.Curve.Curve._mpi_point_from_slicer_buffer (/home/zapu/pgp/kbpgp/lib/ecc/curves.js:64:15)
>
> again, this weird payload works fine in GPG (notice the
> AAAAAAAAAAAAAAAAAA... in armored msg) :
>
> ~/pgp » gpg2
> gpg: Go ahead and type your message ...
> -----BEGIN PGP MESSAGE-----
> Version: Keybase OpenPGP v2.0.58
> Comment: https://keybase.io/crypto
>
> wcA0A6pi0WoSlxTXEgWzBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/ZTuFZa3
> go5bAv8SLZd5vTQzjQiqiXfaQUX3dQu+zytgEeiugIshlJ7JykTPGhdQFVSiKQYe
> a4RKpgbn8SkNhyIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaR1kL+5vnf9E0wv
> vGDKg1jo0uEGYVodxTwk8QuqbxWiCT/jOH9kybjECSPlkDkbUIOezOfaTlwde0Wo
> XUNWZ/WrMOJHerpQvMvPNjEwMSGnsIZPh8/Hafj7j8OauMG5EWgCrzsxv1mgsRXP
> QkNog/5dM9JIAcT8RpDaFecdhRag6ZPuRKmNuhiFtR7o0spcqX2UkJ3FPB7UydX3
> ch9PkTNL1BVD++JqYQE9eaIqlCTAsHwCgO6pQkQqPUvB
> =y7cW
> -----END PGP MESSAGE-----
>
> gpg: encrypted with 521-bit ECDH key, ID AA62D16A129714D7, created 2016-10-15
>       "nist key tester <m+testing@zapu.net>"
> purpleschala
> ^C
> gpg: signal Interrupt caught ... exiting
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keybase/go-crypto/pull/40#issuecomment-275399401>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA05_4iLYwgpnTscLTeNxHbUhNbXmIBJks5rWKwAgaJpZM4LHSlG>
> .
>

zapu commented:
Encrypted messages generated with my go-crypto ecdh code will probably fail in similar way because they round to the number of bits (or rather they receive a buffer and save it with length of `len(buffer)*8`). I'll try to dig up some standard and investigate this more. But KBPGP should probably emulate GPG in this case.

zapu committed:
Reformat comment about MPI padding in ecdh_test

zapu committed:
More style fixes and documentation improvements

zapu commented:
Does it look right here, though? If so, should I make a fix for KBPGP?

zapu committed:
More "return" style changes

maxtaco commented on openpgp/packet/packet.go:
```diff
@@ -557,16 +557,22 @@ func writePointMPI(w io.Writer, curve elliptic.Curve, X, Y *big.Int) (err error)
-	WritePaddedBigInt(w, byteLen, Y)
+	_, err = WritePaddedBigInt(w, byteLen, X)
+	if err != nil {
+		return
```
`return err`

maxtaco commented on openpgp/packet/packet.go:
```diff
@@ -557,16 +557,22 @@ func writePointMPI(w io.Writer, curve elliptic.Curve, X, Y *big.Int) (err error)
+	if err != nil {
+		return
+	}
+	_, err = WritePaddedBigInt(w, byteLen, Y)
```
`return err` just below.

maxtaco commented on openpgp/packet/packet.go:
```diff
@@ -525,6 +526,76 @@ func writeMPI(w io.Writer, bitLength uint16, mpiBytes []byte) (err error) {
+		_, err = readFull(r, Y.bytes)
+		if err != nil {
+			return nil, nil, err
+		}
```
On last thought --- is it important to check that (X,Y) is on the curve here?  We do in the kbpgp implementation.

maxtaco replied:
We don't the curve here, so maybe we should check from the caller?  See comment below.

maxtaco commented on openpgp/packet/encrypted_key.go:
```diff
@@ -48,9 +49,29 @@ func (e *EncryptedKey) parse(r io.Reader) (err error) {
+			return err
+		}
+		e.ecdh_C = make([]byte, int(buf[0]))
+		_, err = readFull(r, e.ecdh_C)
```
Shouldn't we check that (x,y) is on the curve here?

maxtaco commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+func (e *PrivateKey) DecryptShared(X, Y *big.Int) []byte {
+	Sx, _ := e.Curve.ScalarMult(X, Y, e.X.Bytes())
```
Seems like this is right place to check if (X,Y) is on `e.Curve`

maxtaco reviewed (approved):
Ah, I see! This happens by default in `curve.Unmarhsal`. Cheers to the Go folks for not admitting bad data into the program.

maxtaco commented:
Well, it wouldn't be buggy *now* if we didn't check for `IsOnCurve` in `DecryptShared`.  **But**, what if some future version of `elliptic` decides to drop the check?  It might be worth an abundance of caution here and just check it twice.

zapu commented:
But Unmarshal is not used for reading encryption key; only for public key. That being said, I think we could grab entire MPI buffer in `(e *EncryptedKey) parse` instead of decoding numbers; and then use Unmarshal somewhere where we have the curve. This way we could remove readPointMPI (which was introduced by this PR).

maxtaco commented:
The attacks I've read about use a [malicious public key](http://crypto.stackexchange.com/questions/3820/why-do-public-keys-need-to-be-validated).

zapu commented:
Oh, I didn't disagree about the need for this check. I was going through the curve checking before (when I left the `// Note: Unmarshal already checks if point is on curve.` comment in `public_key.go`). Adding more checks sounds like a good idea, though.

So what do you think about getting rid of readPointMPI in favor of less flexible Unmarshal? I'm big fan of having less code :)

maxtaco commented:
How will you read `ecdh_C` in `EncryptedKey#parse` without parsing the two MPIs?

zapu commented:
The point buffer starts with 2-byte integer which is a bit-size of the rest of the buffer; so it's not needed to parse actual numbers to know how long it will be. I think that's how the public key is read - `func readMPI` in `packet.go` seems to read the size in bits and then read rest of the bytes; actual Unmarshaling is a bit later.

zapu commented:
It's a bit confusing, there is nothing about multi-precision integers in `readMPI` to justify the function name. I would need to take a better look, maybe I'm wrong about how this works.

maxtaco commented:
Ok, great. Then I favor this method if at all possible.  If we're always using `curve.Unmarshal`, then it's OK to not check for the point being on the curve explicitly.

zapu commented:
The point-on-curve checking is documented: https://golang.org/pkg/crypto/elliptic/#Unmarshal so hopefully this will not change.

maxtaco commented:
OK, great

zapu committed:
Use ecdh.Unmarshal for encrypted key decoding, remove readPointMPI

zapu commented:
This pushed back the effort to be compatible with GnuPG loose MPI deserializing. So maybe KBPGP should behave in the same way? It would be strict about the total number of bytes for encoded MPI, but not to number of bits (which is not very useful since we are always byte-aligned).

maxtaco commented:
@zapu good with me. Should I merge?

zapu commented:
I'll take another good look through the changes and let you know, ok? Thank you.

PS. Should we copy similar Unmarshal behavior to KBPGP, for this change: https://github.com/keybase/kbpgp/pull/134 ?

zapu committed:
Make linter happy

zapu commented:
I think this looks ready. go-crypt -> KBPGP will faill with things like `Error: Need 263 bits for this curve; got 264` though, but go-crypt -> GnuPG works fine. KBPGP -> go-crypt does as well. There are few things I'm not happy with but let's leave it for future, it's been forever already :)

maxtaco commented:
Will it fail in all cases, or only with weird keys? And are they weird keys or just unlikely keys?

Yeah, sorry to be slow here, can you describe in more detail the case that won't work?  Where does the key originate?  Does it have to be imported/exported? Will it be uploaded to the Website, etc?

zapu commented:
All ECDH messages encrypted with go-crypto would fail in KBPGP (but they work in GnuPG). It's because of that "3 bits for 0x4 header" or "7 bits for 0x40 header" thing.

I'm not sure how to save exact value there (like GnuPG does or KBPGP does) without changing elliptic.Marshal. Marshal returns a byte buffer so I serialize this as `[ len(buffer), buffer... ]`. If Marshal returned `(buf []byte, bitSize int)`, we could improve this.

maxtaco commented:
And does keybase/kbpgp#134 fix this problem?

zapu commented:
Yes, using code from that PR makes it work.

But I'm not sure if allowing arbitrarily sized MPIs is way to go there, I'm wondering if there's some clever attack based on that. Maybe a DoS if you repeatedly pass really huge numbers to some service.

maxtaco commented:
We can obviously put some bound on it if we're afraid of DoS

maxtaco commented:
Let's merge and deploy keybase/kbpgp#134 first then

zapu commented:
Sure, let me put few more hours into it though. Thanks!

zapu commented:
What do you think about something like this: https://gist.github.com/zapu/c60fe9400d79403226ffaf72c3565314

I just realized we don't even need to count bits, because it will always start with 0x4 or 0x40 header. Resulting payloads are readable by current KBPGP master.

maxtaco commented:
👍

zapu committed:
Add bit counting in Marshal to do proper size headers

zapu commented on openpgp/ecdh/ecdh.go:
```diff
@@ -0,0 +1,282 @@
+		val := buffer[0]
+		headerLen = 0
+		for val > 0 {
+			val = val / 2
```
Should we do the bit counting here or just assume 8?

maxtaco replied:
this is great.

zapu replied:
Would something like this be better golang?
```
// Unexpected header - but we can still count the bits.
headerLen = 0
for val := buffer[0]; val > 0; val = val / 2 {
	headerLen++
}
```

maxtaco replied:
I'm indifferent between those two options.

zapu commented:
@maxtaco do you happen to remember the EdDSA issue in KBPGP that was reported? I have an alternative branch for KBPGP that only accept coordinates of good length, but is not strict to the number of bits.

The standard says that the bitSize header has to be set correctly:
```
   The length field of an MPI describes the length starting from its
   most significant non-zero bit.  Thus, the MPI [00 02 01] is not
   formed correctly.  It should be [00 01 01].
```

https://tools.ietf.org/html/rfc4880#section-3.2

And also that the coordinates for ECDH should have a specific length and no other:
```
      B = 04 || x || y

   where x and y are coordinates of the point P = (x, y), each encoded
   in the big-endian format and zero-padded to the adjusted underlying
   field size.  The adjusted underlying field size is the underlying
   field size that is rounded up to the nearest 8-bit boundary.

   Therefore, the exact size of the MPI payload is 515 bits for "Curve
   P-256", 771 for "Curve P-384", and 1059 for "Curve P-521".
```

https://tools.ietf.org/html/rfc6637#section-6

So this pull request sort of follows it but allows for invalid header that's rounded up to the number of bytes. My KBPGP code is here: https://github.com/zapu/kbpgp/blob/8915ff84cc67238254860fc0ca9fbff805b52723/src/ecc/curves.iced#L55-L68

maxtaco commented:
Sadly can't find that issue, I checked a bit in `keybase/clients` and `keybase/keybase-issues`, but the simple queries I though of didn't yield anything. Hmf.

zapu commented:
Almost 3 month-long journey :) thank you for the merge

maxtaco commented:
Great work @zapu! Thanks for this excellent PR.
